### PR TITLE
Add D3D12 test failed callback for pixel history

### PR DESCRIPTION
**Pull request by skarolewics**
This callback determines which tests passed or failed for a given draw during pixel history. The set of tests mirrors the Vulkan implementation, with some adjustments for D3D12 capabilities. The code for updating PixelModification structs during pixel history is included as well since they're related.

### CODE REVIEW

File: renderdoc/driver/d3d12/d3d12_pixelhistory.cpp [hunk 1]
```diff
@@ -121,6 +121,37 @@ struct D3D12PipelineReplacements
...
+  TestEnabled_Scissor = 1 << 2,    // Not valid for D3D12
+  TestEnabled_SampleMask = 1 << 3,
+  TestEnabled_DepthBounds = 1 << 4,
+  TestEnabled_StencilTesting = 1 << 5,
+  TestEnabled_DepthTesting = 1 << 6,
+  TestEnabled_FragmentDiscard = 1 << 7,
+
+  Blending_Enabled = 1 << 8,
+  UnboundFragmentShader = 1 << 9,
+  // TestMustFail_Culling = 1 << 10, // Not valid for D3D12
```
Comment (Zorro666) [inline renderdoc/driver/d3d12/d3d12_pixelhistory.cpp]: I would be tempted to remove this if it its never used or required.

Could add a comment before the enum mentioning it isn't valid compared to vulkan but I don't think that is required (API differences are expected)

File: renderdoc/driver/d3d12/d3d12_pixelhistory.cpp [hunk 2]
```diff
@@ -121,6 +121,37 @@ struct D3D12PipelineReplacements
   ID3D12PipelineState *originalShaderStencil;
 };
 
+enum D3D12PixelHistoryTests : uint32_t
+{
+  TestEnabled_DepthClipping = 1 << 0,
+  TestEnabled_Culling = 1 << 1,
+  TestEnabled_Scissor = 1 << 2,    // Not valid for D3D12
```
Comment (Zorro666) [inline renderdoc/driver/d3d12/d3d12_pixelhistory.cpp]: I was trying to work out if there was an easy way to remove this enum value and simplify the logic.

It does work quite nicely to keep it and have it unconditionally set

File: renderdoc/driver/d3d12/d3d12_pixelhistory.cpp [hunk 3]
```diff
@@ -1107,6 +1138,613 @@ struct D3D12ColorAndStencilCallback : public D3D12PixelHistoryCallback
...
+  if((eventFlags & (TestEnabled_Scissor | TestMustPass_Scissor | TestMustFail_Scissor)) ==
+     TestEnabled_Scissor)
+  {
+    uint64_t occlData = tfCb->GetOcclusionResult(eventId, TestEnabled_Scissor);
+    mod.scissorClipped = (occlData == 0);
+  }
+  if(mod.scissorClipped)
+    return;
+
+  // TODO: Exclusive Scissor Test if NV extension is turned on?
```
Comment (Zorro666) [inline renderdoc/driver/d3d12/d3d12_pixelhistory.cpp:1695]: Does this apply to D3D12 or just a copy'n'paste from Vulkan implementation?

**Review by Zorro666 — COMMENTED** [review]
Thank you - all looks very good.

A couple of very minor queries

File: renderdoc/driver/d3d12/d3d12_pixelhistory.cpp (hunk 3 above)
Comment (skarolewics) [reply renderdoc/driver/d3d12/d3d12_pixelhistory.cpp:1695]: Copy and paste from Vulkan, though I'm not familiar with NVAPI to know if it is possible with D3D12

**Comment by Zorro666** [comment]
Thank you again for the code/feature and making small tweaks during the landing process

**Pull request merged**